packages/core/src/utils/security/handlers.ts

```diff
@@ -0,0 +1,130 @@
+import { Dictionary } from '@stoplight/types/dist';
+import * as Either from 'fp-ts/lib/Either';
+import { fromNullable, getOrElse, mapNullable } from 'fp-ts/lib/Option';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { get } from 'lodash';
+
+type Headers = Dictionary<string, string>;
+
+export type AuthErr = { name: string; message: string; status: number; headers: Headers };
+export type SecurityScheme = { type: string; name: string; in?: string; scheme?: string };
+
+const genUnauthorisedErr = (msg: string): AuthErr => ({
+  name: 'Unauthorised',
+  message: 'Invalid security scheme used',
+  status: 401,
+  headers: { 'WWW-Authenticate': msg },
+});
+
+function isBasicToken(token: string) {
+  const tokenParts = Buffer.from(token, 'base64')
+    .toString()
+    .split(':');
+
+  return tokenParts.length === 2;
+}
+
+const httpBasic = {
+  test: ({ scheme, type }: SecurityScheme) => scheme === 'basic' && type === 'http',
+  handle: <R, I>(someInput: I, name: string, resource?: R) => {
+    const authorizationHeader = get(someInput, ['headers', 'authorization'], '');
+
+    return authorizationHeader
+      ? checkHeader<R>(authorizationHeader, resource)
+      : Either.left(genUnauthorisedErr('Basic realm="*"'));
+  },
+};
+
+function checkHeader<R>(authorizationHeader: string, resource?: R) {
+  const [schema, token] = authorizationHeader.split(' ');
+
+  const isBasicTokenGiven = token && isBasicToken(token);
+  const isBasicSchema = schema === 'Basic';
```

Comment:
The variable name here should probably be `authScheme` just to disambiguate it from a lot of other schemes (and its not a schema). 

We should make it case insensitive too as I checked the HTTP Basic RFC and they mention this:
 
> Note that both scheme and parameter names are matched case-insensitively.

https://tools.ietf.org/html/rfc7617

Make sure the code being written for specific auth types match what the spec says is ok:

https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml

Reply:
@philsturgeon, are all schemes from https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml to be supported?

Reply:
No don't worry, I don't know what many of them are. We have only discussed Basic, Digest, and Bearer.

Reply:
Digest isn't supported for now. Didn't implement it since it isn't (?) openapi supported. Will add it :+1:

Reply:
OpenAPI "supports" anything in the http-authschemes.xhtml list through type === http, but we dont have to write tests for it all. 

The most common are Basic, Digest and Bearer, so having tests for that wouldn't hurt, but if we dont have digest im fine with that. Not important.

Reply:
I meant that there is no implementation for `Digest` (it's not handled) and also no tests for it. I will add it. For now, basic / apiKey / bearer (for the three of: scheme: `bearer` / `type`: oauth2/openIdConnect) are supported.